src/sentry/sentry_metrics/indexer/ratelimiters.py

```diff
@@ -0,0 +1,151 @@
+from typing import MutableMapping, Optional, Sequence, Tuple
+
+from sentry import options
+from sentry.ratelimits.sliding_windows import (
+    GrantedQuota,
+    Quota,
+    RedisSlidingWindowRateLimiter,
+    RequestedQuota,
+    Timestamp,
+)
+from sentry.sentry_metrics.configuration import UseCaseKey, get_ingest_config
+from sentry.sentry_metrics.indexer.base import FetchType, FetchTypeExt, KeyCollection, KeyResult
+from sentry.utils import metrics
+
+OrgId = int
+
+
+def _build_quota_key(use_case_id: UseCaseKey, org_id: Optional[OrgId] = None) -> str:
+    use_case_str = {
+        UseCaseKey.PERFORMANCE: "performance",
+        UseCaseKey.RELEASE_HEALTH: "releasehealth",
+    }[use_case_id]
+
+    if org_id is not None:
+        return f"metrics-indexer-{use_case_str}-org-{org_id}"
+    else:
+        return f"metrics-indexer-{use_case_str}-global"
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quotas")
+def _construct_quotas(use_case_id: UseCaseKey) -> Sequence[Quota]:
+    """
+    Construct write limit's quotas based on current sentry options.
+
+    This value can potentially cached globally as long as it is invalidated
+    when sentry.options are.
+    """
+    if use_case_id == UseCaseKey.PERFORMANCE:
+        return [
+            Quota(prefix_override=_build_quota_key(UseCaseKey.PERFORMANCE, None), **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.global")
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.per-org")
+        ]
+    elif use_case_id == UseCaseKey.RELEASE_HEALTH:
+        return [
+            Quota(prefix_override=_build_quota_key(UseCaseKey.RELEASE_HEALTH, None), **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.global")
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
+def _construct_quota_requests(
+    use_case_id: UseCaseKey, keys: KeyCollection
+) -> Tuple[Sequence[OrgId], Sequence[RequestedQuota]]:
```

Comment:
Why not returning a mapping? It seems it would be a lot more readable.

Reply:
this was already a topic last time I attempted to merge this patchset into master, I wanted to avoid unnecessary type conversions between dict and list.